### ISSUE TITLE
Weird gradient clearing behaviour and erroneous printing

### ISSUE THREAD

**Issue by ifsheldon** [taichi-dev/taichi#2404]
**Describe the bug**
Seems the setting `with ti.Tape(U, clear_gradients = True)` will clear gradients too early. And when I tried to print out gradients, somehow gradients are printed twice even if I called the printing kernel only once.

**To Reproduce**
The code is straightforward and I think you can reproduce the issue with different parameters in the code.
```python
import taichi as ti
import numpy as np
import taichi_glsl as tl

ti.init(arch=ti.cuda)
np.random.seed(2)

transfer_function_len = 5
value_dimension = 1
target_transfer_function = ti.Vector.field(value_dimension, dtype=ti.float32, shape=transfer_function_len)
transfer_function = ti.Vector.field(value_dimension, dtype=ti.float32, shape=transfer_function_len,
                                    needs_grad=True)
sample_num = 10
sample_tex_coords = ti.field(ti.f32, shape=sample_num)
target_samples = ti.Vector.field(transfer_function_len, dtype=ti.float32, shape=sample_num)
samples = ti.Vector.field(value_dimension, dtype=ti.float32, shape=sample_num)
l2loss = ti.field(ti.float32, (), needs_grad=True)

target_transfer_function_np = np.random.rand(transfer_function_len, value_dimension)
target_transfer_function.from_numpy(target_transfer_function_np)


@ti.func
def low_high_frac(x: float):
    low = int(x)
    high = low + 1
    frac = x - low
    return low, high, frac


@ti.func
def sample_transfer_function(scalar: float):
    length = ti.static(target_transfer_function.shape[0])
    val = length * scalar
    low, high, frac = low_high_frac(val)
    val_low = target_transfer_function[low]
    val_high = target_transfer_function[high]
    return tl.mix(val_low, val_high, frac)


@ti.func
def sample_transfer_function_diff(scalar: float):
    length = ti.static(transfer_function.shape[0])
    val = length * scalar
    low, high, frac = low_high_frac(val)
    val_low = transfer_function[low]
    val_high = transfer_function[high]
    return tl.mix(val_low, val_high, frac)


@ti.kernel
def sample_kernel():
    for i in sample_tex_coords:
        tex_coord = sample_tex_coords[i]
        target_value = sample_transfer_function(tex_coord)
        value = sample_transfer_function_diff(tex_coord)
        l2loss[None] += (target_value - value).norm_sqr()


@ti.kernel
def print_clear_grads():
    for i in transfer_function:
        print(i)
        print(transfer_function.grad[i].x)
        # print(transfer_function.grad[i].y)
        # print(transfer_function.grad[i].z)
        transfer_function.grad[i].fill(0.0)


for i in range(2):
    random_tex_coords = np.random.rand(sample_num)
    sample_tex_coords.from_numpy(random_tex_coords)
    sampled_indices = random_tex_coords * transfer_function_len
    sampled_indices_low = sampled_indices.astype(np.int)
    sampled_indices_high = sampled_indices_low + 1
    sampled_indices = np.stack([sampled_indices_low, sampled_indices_high], axis=1)
    print(f"Target transfer function = {target_transfer_function_np}")
    print(f"Epoch {i} Transfer function = {transfer_function.to_numpy()}")
    print(f"Epoch {i} Sampled indices = {sampled_indices}")
    print(f"Epoch {i} Transfer function at {np.unique(sampled_indices)} should have gradients")
    with ti.Tape(l2loss, clear_gradients=False):
        sample_kernel()
        print("Gradients:")
        print_clear_grads()
    l2loss[None] = 0.0
    print("-----------------------------\n")

```

**Log/Screenshots**
If I set `clear_gradients = True`, I got this. It seems gradients are cleared too early, just right after the execution of `sample_kernel` instead of the end of the Tape context, which is weird to me.
```
[Taichi] mode=release
[Taichi] preparing sandbox at /tmp/taichi-uzy5xp2g
[Taichi] version 0.7.20, llvm 10.0.0, commit 284f75ed, linux, python 3.8.8
[TaiGLSL] version 0.0.11
[Taichi] Starting on arch=cuda
[Taichi] materializing...
Target transfer function = [[0.4359949 ]
 [0.02592623]
 [0.54966248]
 [0.43532239]
 [0.4203678 ]]
Epoch 0 Transfer function = [[0.]
 [0.]
 [0.]
 [0.]
 [0.]]
Epoch 0 Sampled indices = [[1 2]
 [1 2]
 [3 4]
 [1 2]
 [1 2]
 [3 4]
 [2 3]
 [0 1]
 [2 3]
 [0 1]]
Epoch 0 Transfer function at [0 1 2 3 4] should have gradients
Gradients:
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
-----------------------------

Target transfer function = [[0.4359949 ]
 [0.02592623]
 [0.54966248]
 [0.43532239]
 [0.4203678 ]]
Epoch 1 Transfer function = [[0.]
 [0.]
 [0.]
 [0.]
 [0.]]
Epoch 1 Sampled indices = [[3 4]
 [4 5]
 [2 3]
 [4 5]
 [0 1]
 [2 3]
 [0 1]
 [2 3]
 [0 1]
 [0 1]]
Epoch 1 Transfer function at [0 1 2 3 4 5] should have gradients
Gradients:
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
-----------------------------

```

If I set `clear_gradients=False`, I got this.
```
[Taichi] mode=release
[Taichi] preparing sandbox at /tmp/taichi-hjluoxas
[Taichi] version 0.7.20, llvm 10.0.0, commit 284f75ed, linux, python 3.8.8
[TaiGLSL] version 0.0.11
[Taichi] Starting on arch=cuda
[Taichi] materializing...
Target transfer function = [[0.4359949 ]
 [0.02592623]
 [0.54966248]
 [0.43532239]
 [0.4203678 ]]
Epoch 0 Transfer function = [[0.]
 [0.]
 [0.]
 [0.]
 [0.]]
Epoch 0 Sampled indices = [[1 2]
 [1 2]
 [3 4]
 [1 2]
 [1 2]
 [3 4]
 [2 3]
 [0 1]
 [2 3]
 [0 1]]
Epoch 0 Transfer function at [0 1 2 3 4] should have gradients
Gradients:
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
-----------------------------

Target transfer function = [[0.4359949 ]
 [0.02592623]
 [0.54966248]
 [0.43532239]
 [0.4203678 ]]
Epoch 1 Transfer function = [[0.]
 [0.]
 [0.]
 [0.]
 [0.]]
Epoch 1 Sampled indices = [[3 4]
 [4 5]
 [2 3]
 [4 5]
 [0 1]
 [2 3]
 [0 1]
 [2 3]
 [0 1]
 [0 1]]
Epoch 1 Transfer function at [0 1 2 3 4 5] should have gradients
Gradients:
0
1
2
3
4
-0.113710
-1.207789
-1.656651
-2.725014
-0.175279
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
-----------------------------

```

This is even more weird to me for two reasons:
1. I should have gradients at Epoch 0, but I got all zeros. However, somehow, in Epoch 1, I got gradients, which seems working correctly.
2. In Epoch 1, you can see indices and gradients are printed twice, but I only call `print_clear_grads` once per epoch.
3. Even more interestingly, the two printing results of Epoch 1 are not the same. The first set of printed gradients is non-zero, but the second one is.

**Comment by k-ye** [comment]
I don't have a laptop at hand. What happens if you remove 'transfer_function.grad[i].fill(0.0)'?

**Comment by k-ye** [comment]
I don't see why this is erroneous. In the forward step, you've cleared the gradients manually in `print_clear_grads` in the last step, after the gradients are computed. Just rely on the `clear_gradients=True` to clear the gradients for you.

**Comment by ifsheldon** [comment]
Seems you misunderstood the code. In the code, I call `sample_kernel()` first and then `print_clear_grads()` per epoch, so I expect the printed gradients are non-zero, but the printed gradients of the first epoch is 0.

The printing behavior is weird in that for each epoch, the gradients are printed twice, but I only call `print_clear_grads()` once per epoch. Moreover, the two versions of printed gradients are different in the second epoch.

**Comment by ifsheldon** [comment]
See the code here.
```python
  with ti.Tape(l2loss, clear_gradients=True):
      sample_kernel()
      print("Gradients:")
      print_clear_grads()
  # expect gradient clear here
```
As stated before, `sample_kernel()` and `print_clear_grads()` are both in the context of `ti.Tape`, so I expect the gradients are clear automatically after calling `print_clear_grads()` (although I cleared them in `print_clear_grads()`), but the printed gradients are zeros.
Let's make things clearer and change the code like below.
```python
@ti.kernel
def print_grads(): # changed
    for i in transfer_function:
        print(i)
        print(transfer_function.grad[i].x)
        # print(transfer_function.grad[i].y)
        # print(transfer_function.grad[i].z)
        # transfer_function.grad[i].fill(0.0) # changed here


for i in range(2):
    random_tex_coords = np.random.rand(sample_num)
    sample_tex_coords.from_numpy(random_tex_coords)
    sampled_indices = random_tex_coords * transfer_function_len
    sampled_indices_low = sampled_indices.astype(np.int)
    sampled_indices_high = sampled_indices_low + 1
    sampled_indices = np.stack([sampled_indices_low, sampled_indices_high], axis=1)
    print(f"Target transfer function = {target_transfer_function_np}")
    print(f"Epoch {i} Transfer function = {transfer_function.to_numpy()}")
    print(f"Epoch {i} Sampled indices = {sampled_indices}")
    print(f"Epoch {i} Transfer function at {np.unique(sampled_indices)} should have gradients")
    with ti.Tape(l2loss, clear_gradients=True):
        sample_kernel()
        print("Gradients:")
        print_grads() # changed
    l2loss[None] = 0.0
    print("-----------------------------\n")
```
Now you will get the output like below:
```
[Taichi] mode=release
[Taichi] preparing sandbox at /tmp/taichi-fi420133
[Taichi] version 0.7.20, llvm 10.0.0, commit 284f75ed, linux, python 3.8.8
[TaiGLSL] version 0.0.11
[Taichi] Starting on arch=cuda
[Taichi] materializing...
Target transfer function = [[0.4359949 ]
 [0.02592623]
 [0.54966248]
 [0.43532239]
 [0.4203678 ]]
Epoch 0 Transfer function = [[0.]
 [0.]
 [0.]
 [0.]
 [0.]]
Epoch 0 Sampled indices = [[1 2]
 [1 2]
 [3 4]
 [1 2]
 [1 2]
 [3 4]
 [2 3]
 [0 1]
 [2 3]
 [0 1]]
Epoch 0 Transfer function at [0 1 2 3 4] should have gradients
Gradients:
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
-----------------------------

Target transfer function = [[0.4359949 ]
 [0.02592623]
 [0.54966248]
 [0.43532239]
 [0.4203678 ]]
Epoch 1 Transfer function = [[0.]
 [0.]
 [0.]
 [0.]
 [0.]]
Epoch 1 Sampled indices = [[3 4]
 [4 5]
 [2 3]
 [4 5]
 [0 1]
 [2 3]
 [0 1]
 [2 3]
 [0 1]
 [0 1]]
Epoch 1 Transfer function at [0 1 2 3 4 5] should have gradients
Gradients:
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
0
1
2
3
4
0.000000
0.000000
0.000000
0.000000
0.000000
-----------------------------
```
which makes no sense.

**Comment by ifsheldon** [comment]
> In the forward step, you've cleared the gradients manually in print_clear_grads in the last step, after the gradients are computed
The problem is before clearing gradients, I print them first, but I got incorrect gradients.
> Just rely on the clear_gradients=True to clear the gradients for you.
It seems to clear gradients prematurely, which is what I meant by "weird gradient clearing".

**Comment by ifsheldon** [comment]
@k-ye Any suggestions?

**Comment by k-ye** [comment]
Hey @ifsheldon, sorry i just got some time to reply.

Here's roughly how `ti.Tape` works:  It runs the kernels in the forward pass, records all the launched kernels, then **runs them in the backward pass automatically and implicitly** to compute the gradients. For each kernel run in the forward pass, Taichi automatically generates the corresponding adjoint kernel computing the gradients. This is just how most reverse-mode AD algorithms are implemented. There are two implications from this:

1. `print_grads()` gets called twice.
2. Gradients are not computed until the backward pass is done. This is why all the printed gradients are zero.

As for `clear_gradients=True`, this will clear all the gradients **before** running the forward pass, not after the entire context. See https://github.com/taichi-dev/taichi/blob/15465fb6c642e5fefee0c755e5c9e330f8964305/python/taichi/lang/__init__.py#L364-L374

Try simply moving `print_grads()` out of `ti.Tape`, and you should be able to observe the correct gradients:

```py
with ti.Tape(l2loss, clear_gradients=False):
    sample_kernel()

print("Gradients:")
print_clear_grads()
```

Lastly, since you are working heavily with Taichi, please feel free to add more documentation if you find it not clear. We welcome all sorts of contribution, really appreciated!

**Comment by ifsheldon** [comment]
@k-ye OK, I see, thanks a lot. It's somewhat like the GradientTape in TensorFlow, but I'm not very familiar with TensorFlow. I actually took a look at the code you referred when I tried to figure out what's going on, so I was confused when I see `if clear_gradients` branch, because I got used to the paradigm Pytorch adopts. 

Something like `U.backward()` is more intuitive for me, which is just a minor API suggestion for you to consider because I don't know if you have other reasons to use `ti.Tape`. Maybe @yuanming-hu may explain a bit on why `ti.Tape` or comment a bit on my suggestion?

From what I understand here, there's an implicit  "`backward()`" called when the context `ti.Tape` ends. It's somewhat like
```python
with ti.Tape(l2loss, clear_gradients=False):
    sample_kernel()
    print_grad()
    backward() # implicitly
```
This may lead to confusions that I have in this issue. The way I suggest is somewhat like
```python
sample_kernel()
l2loss.backward() # this reminds us we need to do backward first then print gradients
print_grad()
```
But anyway, clearly, the documentation in the section of Differentiable programming is not enough and I will try to add some content to that after I finished my differentiable DVR renderer. 

BTW, I really need your help in the [post](https://forum.taichi.graphics/t/differentiable-direct-volume-renderer-gradient-issues/1496) in the forum, which is also about gradients. Please take a look at it.
